### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by krl**
Testing + multiaddr support.

The config replace still fails, which seems to be the case also with the cli command, i'll assume it's a problem in the backend.

### CODE REVIEW

**Comment by jbenet** [comment]
> The config replace still fails

what does this mean? expand?

(also should this fail fail the travis tests? or are we ignoring it?)

File: index.js [hunk 1]
```diff
@@ -13,10 +13,19 @@ try {
...
-  if (!host) host = 'localhost'
+module.exports = function (host_or_multiaddr, port) {
+  var host
+
+  if (!host_or_multiaddr) host = 'localhost'
   if (!port) port = 5001
 
+  if (host_or_multiaddr.match('/')) {
+    // is a multiaddr
+    var parts = host_or_multiaddr.split('/')
```
Comment (travisperson) [inline index.js]: Could probably use the multiaddr package here instead.

https://github.com/jbenet/node-multiaddr

**Comment by travisperson** [comment]
Ping

**Comment by krl** [comment]
pong, using multiaddr, leaving config/replace commented out for now until i can actually figure out what's going on.

**Comment by travisperson** [comment]
Okay cool, I just want to get the testing merged :smile:

**Comment by travisperson** [comment]
Just a note: I think we also want to include mocha as a dev dependency.

**Comment by harlantwood** [comment]
Yes, and also add the test command to package.json, so that `npm test` works.  I'd be happy to add a CI config file if you like BTW... I love and recommend circleci, and of course Travis is fine too.

**Comment by travisperson** [comment]
I believe we are moving to circleci, or at least it's been floated around as Travis is queueing up a lot and I believe we can get more throughput with circleci. I don't know much though.

Either way a config would be great.

/cc @jbenet

**Comment by jbenet** [comment]
> I believe we are moving to circleci, or at least it's been floated around as Travis is queueing up a lot and I believe we can get more throughput with circleci. I don't know much though.

Travis is per-repo so making other repos with travis is fine. making it use circle-ci too is fine as well. there's a bunch of cli tools that make it easy to manipulate travis, not sure about circle-ci. circle-ci is glorious (beautiful, fast, etc). not sure about OSS model? i think it's free but not sure if it's as nice + committed as travis. (travis tends to always just work, even if slow.

speed is not something to be concerned with here i dont think. go-ipfs is really slow but those tests are beasts.

can also split up into multiple repos (that will help go-ipfs later).

---

Is this merge ready?

**Comment by travisperson** [comment]
>Travis is per-repo so making other repos with travis is fine.

Perfect, I think we can definitely use that unless there is a completing reason not too. I have yet to really look at circleci, so nor aware of any amazing features it may have.

I think it will be as soon as we get the mocha dev dependency. All tests pass!

**Comment by harlantwood** [comment]
> there's a bunch of cli tools that make it easy to manipulate travis, not sure about circle-ci. 

Hm, the CLI tools for Circle seem quite immature, in a quick search.  This may be an area where Travis wins (for now).

> not sure about OSS model? i think it's free but not sure if it's as nice + committed as travis. 

They are quite supportive of OSS:
>  We’ve made that first container free, while additional containers still cost $50/month. 
> ### Open source
> If you use CircleCI for your open source projects, this doesn’t affect you: **open source projects still get 3 free containers**.

@travisperson it's not that Circle is so amazing per say, it's just really really nice to use.  As I said on the topic in ipfs/infrastructure#20:

> I used Travis before Circle, and I'd never go back. I can't point to features Circle has that Travis doesn't; but everything feels even cleaner, more lucid, more usable. They are always adding new features, and are eager to hear what users want.

Maybe we should move any further discussion on CI over to ipfs/infrastructure#20 as it is generally useful across multiple IPFS projects.

**Comment by harlantwood** [comment]
I realized BTW that no Circle CI config would be needed -- once `npm test` works, we just need a project adminstrator (not just collaborator) to visit https://circleci.com/gh/ipfs/node-ipfs-api/edit and click the `Follow` button. That's it.

**Comment by jbenet** [comment]
@harlantwood cool! let's try it out. we could run both travis and circle and see how it goes.

**Comment by harlantwood** [comment]
Sure!

**Comment by travisperson** [comment]
I'm going to update this to add `mocha`.

**Comment by jbenet** [comment]
@travisperson have the scripts be `npm test` as other things like to depend on that

**Comment by jbenet** [comment]
thanks @travisperson @krl

**Comment by harlantwood** [comment]
To turn on Circle, we just need a project adminstrator (not just collaborator) to visit https://circleci.com/gh/ipfs/node-ipfs-api/edit and click the `Follow` button.

**Comment by jbenet** [comment]
It's on now! :+1:

**Pull request merged**